libclamav/crypto.c

```diff
@@ -260,7 +348,12 @@ unsigned char *cl_hash_file_fd(int fd, const char *alg, unsigned int *olen)
     const EVP_MD *md;
     unsigned char *res;
 
-    md = EVP_get_digestbyname(alg);
+    // If OpenSSL is running in FIPS mode, we need to use the FIPS-compliant md5 lookup of the algorithm
+    if (cli_fips_mode && !strncasecmp(alg, "md5", 3))
+        md = EVP_MD_fetch(NULL, alg, "-fips");
```

Comment:
I know this PR is much bigger than #959, but this section of code appears to be the main as that PR's. Is this introducing the same issues as that, and would this have the same potential memory leak mentioned on that PR? I didn't see any added calls to EVP_MD_free in this PR.

Reply:
Hi David!  So per the OpenSSL documentation, we don't really need to worry about freeing this.  The first element of the OpenSSL documentation I want to cite is:

> * EVP_MD_free()
> 
> Decrements the reference count for the fetched EVP_MD structure. If the reference count drops to 0 then the structure is freed. If the argument is NULL, nothing is done.

So, we're really dealing with a reference count.  While this means it's not really an issue, I also like to be tidy with reference counts, but that leads to the second section of documentation I wanted to highlight:

> EVP_MD_fetch()
> 
> Fetches the digest implementation for the given algorithm from any provider offering it, within the criteria given by the properties. See ["ALGORITHM FETCHING" in crypto(7)](https://docs.openssl.org/master/man7/ossl-guide-libcrypto-introduction/) for further information.
> 
> The returned value must eventually be freed with EVP_MD_free().
> 
> Fetched EVP_MD structures are reference counted.

So, we *should* free when using EVP_MD_fetch.  However we also have:

> EVP_get_digestbyname(), EVP_get_digestbynid(), EVP_get_digestbyobj()
> 
> Returns an EVP_MD structure when passed a digest name, a digest NID or an ASN1_OBJECT structure respectively.
> 
> The EVP_get_digestbyname() function is present for backwards compatibility with OpenSSL prior to version 3 and is different to the EVP_MD_fetch() function since it does not attempt to "fetch" an implementation of the cipher. Additionally, it only knows about digests that are built-in to OpenSSL and have an associated NID. Similarly EVP_get_digestbynid() and EVP_get_digestbyobj() also return objects without an associated implementation.
> 
> When the digest objects returned by these functions are used (such as in a call to EVP_DigestInit_ex()) an implementation of the digest will be implicitly fetched from the loaded providers. This fetch could fail if no suitable implementation is available. Use EVP_MD_fetch() instead to explicitly fetch the algorithm and an associated implementation from a provider.
> 
> See ["ALGORITHM FETCHING" in crypto(7)](https://docs.openssl.org/master/man7/ossl-guide-libcrypto-introduction/) for more information about fetching.
> 
> The digest objects returned from these functions do not need to be freed with EVP_MD_free().

So, now we do not have to free when using EVP_get_digestbyname.  Terribly confusing.  That said, I've added the logic to free according to the documentation.   Please see the updated commit.

I also added ECC secp521r1 support for NIST Approved Elliptical Curve signing, as well.

Reply:
Thanks! I appreciate the detailed explanation. I agree, that documentation seems inconsistent with itself.